Use square versions of distance/length where appropriate

AThousandShips opened the pull request:
In cases where the actual value is not directly relevant it's unnecessary to take the square root, as squares preserve ordering

Split into three separate commits for ease of review, with increasing complexity of cases

Very open to detailed tweaks in the exact solutions, some cases are a bit messy for readability but I think they're still worth avoiding the square root

The exact limits in some cases can also be discussed to avoid clutter and complexity, but the current solution keeps the exact comparisons from the original taking the square of the compared value

There are a few more cases that could be improved with this where the value is used specifically, I left those out but can add them in a follow-up where we instead take the square root manually at the end (a few cases like that where the closest value is calculated, and then used once at the end. In cases like that we can do the computations on the square and then take the root at the end)

Left a few cases that were especially messy for readability, and can restore some of the explicit expressions in some of the cases if the meaning of the values are more relevant, i.e. `10.0 * 10.0` instead of `100.0`, or more relevant `(5.0 * EDSCALE) * (5.0 * EDSCALE)` over `25.0 * EDSCALE * EDSCALE` etc., simplified most of the expressions for readability but it might be relevant to keep the original value explicitly.

Created temporaries for a lot of cases to help readability, they should simply be optimized out in cases where they're used exactly once, can restore them to just `foo * foo` instead of `foo_squared` but felt they were more readable this way

Considered adding a helper `SQUARE` macro that just does `#define SQUARE(m_arg) ((m_arg) * (m_arg))` but felt it obscured things too much, it would help readability in a lot of cases though assuming it isn't too obscuring
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

AThousandShips commented on editor/animation/animation_state_machine_editor.cpp:
```diff
@@ -252,14 +252,14 @@ void AnimationNodeStateMachineEditor::_state_machine_gui_input(const Ref<InputEv
 		// Use midpoint distance as bias.
 		if (close_candidates.size() > 1) {
-			float best_midpoint_dist = 1e20;
+			float best_midpoint_dist_squared = 1e20;
```
For all the cases like these I left the default value as is, in most cases the square of the actual value isn't representable (`1e40` is smaller than the smallest value representable by `float`)

aaronfranke replied:
Generally the point of these values is to be "effectively infinity" so that all other values are below it, so larger values should generally be better. I wonder if there's any reason to not use infinity itself?

AThousandShips replied:
I'll separate out these cases

AThousandShips commented on editor/scene/3d/node_3d_editor_plugin.cpp:
```diff
@@ -861,20 +862,20 @@ ObjectID Node3DEditorViewport::_select_ray(const Point2 &p_pos) const {
+			const real_t dist_squared = pos.distance_squared_to(point);
 
-			if (dist < 0) {
+			if (dist_squared < 0) {
```
There are a few cases like this across the code base that make no sense, didn't change it in this case but would consider removing them as they are never true

AThousandShips commented on scene/resources/animation.cpp:
```diff
@@ -4154,11 +4154,12 @@ bool Animation::_float_track_optimize_key(const TKey<float> t0, const TKey<float
 		return true;
 	}
-	if ((t0.value - t1.value).length() < p_allowed_precision_error && (t1.value - t2.value).length() < p_allowed_precision_error) {
+	if ((t0.value - t1.value).length_squared() < allowed_precision_error_squared && (t1.value - t2.value).length_squared() < allowed_precision_error_squared) {
```
These can all probably be replaced with:
```suggestion
	if (t0.value.distance_squared_to(t1.value) < allowed_precision_error_squared && t1.valuedistance_squared_to(t2.value) < allowed_precision_error_squared) {
```
But left them as is for now to reduce the changes

lawnjelly commented:
Aside from the promotion issue I mentioned above there's a couple of other aspects (that if I don't bring up someone else will):

* If this is not a bottleneck, it doesn't strictly speaking matter (unless for consistency). In some cases using length directly could be clearer.
* The squared  constant may be less clear in some situations, maybe it could be worth considering using `Math.square(8)` (or some equivalent `constexpr`) in some cases.

aaronfranke commented on scene/3d/physics/character_body_3d.cpp:
```diff
@@ -186,9 +186,9 @@ void CharacterBody3D::_move_and_slide_grounded(double p_delta, bool p_was_on_flo
 			}
 
-			if (collision_state.floor && floor_stop_on_slope && (velocity.normalized() + up_direction).length() < 0.01) {
+			if (collision_state.floor && floor_stop_on_slope && (velocity.normalized() + up_direction).length_squared() < 0.0001) {
```
```suggestion
			if (collision_state.floor && floor_stop_on_slope && (velocity.normalized() + up_direction).length_squared() < 0.0001f) {
```

aaronfranke commented on scene/3d/physics/character_body_3d.cpp:
```diff
@@ -176,7 +176,7 @@ void CharacterBody3D::_move_and_slide_grounded(double p_delta, bool p_was_on_flo
 			if (collision_state.ceiling && platform_ceiling_velocity != Vector3() && platform_ceiling_velocity.dot(up_direction) < 0) {
 				// If ceiling sliding is on, only apply when the ceiling is flat or when the motion is upward.
-				if (!slide_on_ceiling || motion.dot(up_direction) < 0 || (ceiling_normal + up_direction).length() < 0.01) {
+				if (!slide_on_ceiling || motion.dot(up_direction) < 0 || (ceiling_normal + up_direction).length_squared() < 0.0001) {
```
```suggestion
				if (!slide_on_ceiling || motion.dot(up_direction) < 0 || (ceiling_normal + up_direction).length_squared() < 0.0001f) {
```

aaronfranke commented on scene/3d/cpu_particles_3d.cpp:
```diff
@@ -1156,7 +1156,7 @@ void CPUParticles3D::_particles_process(double p_delta) {
 		if (particle_flags[PARTICLE_FLAG_DISABLE_Z]) {
 			if (particle_flags[PARTICLE_FLAG_ALIGN_Y_TO_VELOCITY]) {
-				if (p.velocity.length() > 0.0) {
+				if (p.velocity.length_squared() > 0.0) {
```
Pre-existing problem, but all the changed comparisons in this file should be given an `f` suffix.

```suggestion
				if (p.velocity.length_squared() > 0.0f) {
```

aaronfranke commented on scene/2d/physics/character_body_2d.cpp:
```diff
@@ -149,7 +149,7 @@ void CharacterBody2D::_move_and_slide_grounded(double p_delta, bool p_was_on_flo
 			if (on_ceiling && result.collider_velocity != Vector2() && result.collider_velocity.dot(up_direction) < 0) {
 				// If ceiling sliding is on, only apply when the ceiling is flat or when the motion is upward.
-				if (!slide_on_ceiling || motion.dot(up_direction) < 0 || (result.collision_normal + up_direction).length() < 0.01) {
+				if (!slide_on_ceiling || motion.dot(up_direction) < 0 || (result.collision_normal + up_direction).length_squared() < 0.0001) {
```
```suggestion
				if (!slide_on_ceiling || motion.dot(up_direction) < 0 || (result.collision_normal + up_direction).length_squared() < 0.0001f) {
```

aaronfranke commented on scene/2d/cpu_particles_2d.cpp:
```diff
@@ -1096,7 +1096,7 @@ void CPUParticles2D::_particles_process(double p_delta) {
 
 		if (particle_flags[PARTICLE_FLAG_ALIGN_Y_TO_VELOCITY]) {
-			if (p.velocity.length() > 0.0) {
+			if (p.velocity.length_squared() > 0.0) {
```
Same pre-existing problem, all the changed comparisons in this file should be given an `f` suffix.

```suggestion
			if (p.velocity.length_squared() > 0.0f) {
```

aaronfranke commented on platform/web/display_server_web.cpp:
```diff
@@ -264,7 +264,7 @@ int DisplayServerWeb::_mouse_button_callback(int p_pressed, int p_button, double
 
 		if (ev->get_button_index() == ds->last_click_button_index) {
-			if (diff < 400 && Point2(ds->last_click_pos).distance_to(ev->get_position()) < 5) {
+			if (diff < 400 && Point2(ds->last_click_pos).distance_squared_to(ev->get_position()) < 25.0) {
```
```suggestion
			if (diff < 400 && Point2(ds->last_click_pos).distance_squared_to(ev->get_position()) < 25.0f) {
```

aaronfranke commented on platform/linuxbsd/x11/display_server_x11.cpp:
```diff
@@ -5241,7 +5241,7 @@ void DisplayServerX11::process_events() {
 
 					if (mb->get_button_index() == last_click_button_index) {
-						if (diff < 400 && Vector2(last_click_pos).distance_to(Vector2(event.xbutton.x, event.xbutton.y)) < 5) {
+						if (diff < 400 && Vector2(last_click_pos).distance_squared_to(Vector2(event.xbutton.x, event.xbutton.y)) < 25.0) {
```
```suggestion
						if (diff < 400 && Vector2(last_click_pos).distance_squared_to(Vector2(event.xbutton.x, event.xbutton.y)) < 25.0f) {
```

aaronfranke commented on platform/linuxbsd/wayland/wayland_thread.cpp:
```diff
@@ -2836,7 +2836,7 @@ void WaylandThread::_wp_tablet_tool_on_frame(void *data, struct zwp_tablet_tool_
 				}
 
-				if (old_td.double_click_begun && mb->is_pressed() && td.last_button_pressed == old_td.last_button_pressed && (td.button_time - old_td.button_time) < 400 && Vector2(td.last_pressed_position * scale).distance_to(Vector2(old_td.last_pressed_position * scale)) < 5) {
+				if (old_td.double_click_begun && mb->is_pressed() && td.last_button_pressed == old_td.last_button_pressed && (td.button_time - old_td.button_time) < 400 && Vector2(td.last_pressed_position * scale).distance_squared_to(Vector2(old_td.last_pressed_position * scale)) < 25.0) {
```
```suggestion
				if (old_td.double_click_begun && mb->is_pressed() && td.last_button_pressed == old_td.last_button_pressed && (td.button_time - old_td.button_time) < 400 && Vector2(td.last_pressed_position * scale).distance_squared_to(Vector2(old_td.last_pressed_position * scale)) < 25.0f) {
```

aaronfranke commented on platform/linuxbsd/wayland/wayland_thread.cpp:
```diff
@@ -1893,7 +1893,7 @@ void WaylandThread::_wl_pointer_on_frame(void *data, struct wl_pointer *wl_point
 				}
 
-				if (old_pd.double_click_begun && mb->is_pressed() && pd.last_button_pressed == old_pd.last_button_pressed && (pd.button_time - old_pd.button_time) < 400 && Vector2(old_pd.last_pressed_position * scale).distance_to(Vector2(pd.last_pressed_position * scale)) < 5) {
+				if (old_pd.double_click_begun && mb->is_pressed() && pd.last_button_pressed == old_pd.last_button_pressed && (pd.button_time - old_pd.button_time) < 400 && Vector2(old_pd.last_pressed_position * scale).distance_squared_to(Vector2(pd.last_pressed_position * scale)) < 25.0) {
```
At this point I realized I've reviewed half the PR and have been writing way too many of these comments, I'll stop for now and look over things again. There are more cases of things that need to be `f` suffixed.

```suggestion
				if (old_pd.double_click_begun && mb->is_pressed() && pd.last_button_pressed == old_pd.last_button_pressed && (pd.button_time - old_pd.button_time) < 400 && Vector2(old_pd.last_pressed_position * scale).distance_squared_to(Vector2(pd.last_pressed_position * scale)) < 25.0f) {
```

aaronfranke commented on modules/navigation_3d/editor/navigation_obstacle_3d_editor_plugin.cpp:
```diff
@@ -621,12 +622,12 @@ EditorPlugin::AfterGUIInput NavigationObstacle3DEditorPlugin::forward_3d_gui_inp
 				if (mb->get_button_index() == MouseButton::LEFT && mb->is_pressed()) {
 					int closest_idx = -1;
-					real_t closest_dist = 1e10;
+					real_t closest_dist_squared = 1e10;
```
Keeping the existing behavior would be:

```suggestion
					real_t closest_dist_squared = 1e20;
```

There are more cases in this file.

AThousandShips commented on editor/scene/2d/physics/cast_2d_editor_plugin.cpp:
```diff
@@ -74,7 +74,7 @@ bool Cast2DEditor::forward_canvas_gui_input(const Ref<InputEvent> &p_event) {
 
 		if (mb->is_pressed()) {
-			if (xform.xform(target_position).distance_to(gpoint) < 8) {
+			if (xform.xform(target_position).distance_squared_to(gpoint) < 64.0) {
```
I'll fix those! It's `real_t` so `float` sounds best as it'll promote anyway

AThousandShips commented:
I'll reduce it to obvious cases (plain comparisons with simple constants, and loops) and put the less obvious cases in a separate branch for the future

For that I'll add a `Math::square` as well for ease of use

AThousandShips commented:
GDScript is not useful to measure this, the overhead of GGScript is far higher than any difference in the actual operations

mrjustaguy commented:
True, the overhead of adding an operation is there. however funny enough running 10m operations with just `distance_to` and `distance_squared_to` resulted in consistent 580-600ms for `distance_to` and `distance_squared_to` ended up increasing to 591ms-615ms, and the results are quite consistently being reproduced on my machine.

I don't understand why the more `distance_squared_to` is used it gets more expensive compared to `distance_to` as that makes little sense to me, but it's a consistent phenomenon on my i3-10105f

AThousandShips commented:
But if you use GDScript the process of the VM is orders of magnitude slower than any other operation here, GDScript is not useful to measure here, but it's irrelevant to this PR, might be worth investigating

mrjustaguy commented:
I'm not trying to benchmark how fast the operations themselves are, I'm trying to get how the relative performance of the two acts to a point, which when I'm just calling one or the other from what I understand should just be GDScript overhead of calling a builtin function + the cost of the function, and the overhead can be cancelled out when you're calling the same number of functions on both ends

AThousandShips committed:
Use square versions of distance/length where appropriate

AThousandShips committed:
Second stage

AThousandShips committed:
Third stage

AThousandShips commented:
Reworked the cases:
* Restored some expressions for clarity
* Added some temporaries for clarity
* Fixed literals
* Excluded some loop cases

Kept loop cases where the starting value was:
* `FLT_MAX`, which we can assume is supposed to not matter
* A value that can be squared within the range of the value
* A starting value that's well defined

Kept the rest for future use

Left as three commits for now for ease of review, first one being simple (cases with a single line, that's a literal, or in one case using a constant), second being cases with either a separate temporary, or a product in the expression, and last being loops

Ivorforce commented:
> The thing I was trying to get across is, this needs to be benchmarked pre vs post changes, by stressing the functions changed as the changes are such micro optimizations that they may themselves end up not working as expected, That's all.

Sorry if we're coming across as dismissive. Benchmarking changes is never wrong, so it's a good idea in principle.
In this particular case we are reasonably confident that the change is not risky due to our understanding of operator costs and inlining, with the main concern as @lawnjelky expressed being small effects that may not be worth the churn.

AThousandShips commented:
If you find data on that square roots are somehow faster than multiplications then please share, but I don't think it's relevant to entertain that hypothetical

But all that would mean is removing some of the cases that involve just a local multiplication of a non-constant, which are a minority of the cases

Let's focus on the readability and correctness

mrjustaguy commented:
using Squared values very often comes at the cost of readability (not always mind you) and isn't worth sacrificing for a good practice that may have gone stale (when were these ground truths last tested to be valid? was it recently, or a couple of decades ago or more?)
Also important to note _all hardware that runs Godot_, as maybe the Core 2 Duo would benefit (but as of SSE4.2 it no longer does), but a 3rd gen i7 would not.

AThousandShips commented:
I'm fully open to removing any readability hurting cases (to be deferred to a future where we add `Math::square`, which I'm testing a bit)

It'd be worth also looking at moving other code where relevant, but most modern compilers will fix that (including repeated multiplications of the same value)

fire commented:
Since this is an optimizing pull request I would suggest measuring original baseline vs the current code enhancement to avoid performance regressions.

AThousandShips commented:
What areas are you suggesting? What baseline?

fire commented:
We have a benchmark server so we pick any stable version you want as the base line, the current master and the current master with this pull request.

We should only merge this if there are improvements.

AThousandShips commented:
I know what areas I've changed 🤣 but you suggested benchmarks so I assumed you had an idea of what areas to specifically look for, I do know how to benchmark and was going to do so

fire commented:
> I know what areas I've changed 🤣 but you suggested benchmarks so I assumed you had an idea of what areas to specifically look for, I do know how to benchmark and was going to do so
> 
> Please don't assume that I don't know what I'm doing

To be clear:

I would suggest 1. skeletons 2. physics

AThousandShips commented:
Prepped the benchmarks and will compare the data today, initial look show some improvements but will have to crunch the data
